### PR TITLE
[FLINK-26394] CheckpointCoordinator.isTriggering can not be reset in a particular case.

### PR DESCRIPTION
**Pull request by pltbkd**
## What is the purpose of the change

CheckpointCoordinator.isTriggering can not be reset if a checkpoint expires while the checkpointCoordinator task is queuing in the SourceCoordinator executor. Once the issue happens the job can never recover until restarted. This PR is fixing this issue.

## Brief change log

  - Throw timeout exception if the checkpoint completable future expires. The timeout is same as the checkpoint timeout.

## Verifying this change

Tests will be added later.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by pltbkd** [comment]
This is a preview for discussion. Please do not merge at present.

**Comment by pltbkd** [comment]
> @pltbkd Could you please add a ITcase to reproduce this problem?
> 
> BTW, you can tag this PR as `draft` if you think this PR is still under development.

Thanks for the suggestion and sorry for the late reply.

I have added a UT case for CheckpointCoordinator. It would fail without the patch.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java
Comment (becketqin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:128]: Could the pending future just be a part of `PendingCheckpoint`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java
Comment (becketqin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:673]: It looks that this only happens when the pending checkpoint got aborted right after it is created. Should probably use `PendingCheckpoint.getFailureCause()` instead.

**Review by becketqin — COMMENTED** [review]
@pltbkd  Thanks for the patch. LGTM overall, just two minor comments.

**Comment by pltbkd** [comment]
@becketqin 
Thanks for the review! I've updated the PR fixing these issues.
The future is now added in the constructor of the PendingCheckpoint, and we no longer need to handle the case of the second issue.

**Review by becketqin — APPROVED** [review]
Thanks for the updating the patch. LGTM. Will merge the patch after the CI passes.

**Comment by pltbkd** [comment]
I found the CI has failed and I'm finding out the reason.

**Comment by pltbkd** [comment]
@flinkbot run azure

**Comment by pltbkd** [comment]
@flinkbot run azure

**Comment by pltbkd** [comment]
@flinkbot run azure

**Comment by pltbkd** [comment]
@flinkbot run azure

**Comment by becketqin** [comment]
The CI failure seems irrelevant.  I'll just merge the PR.

**Comment by becketqin** [comment]
Merged.
master: 31222b9adf1c354b22fd50c587efc16734b18d40
release-1.15: 9fc89a05f128ab645b73687f240fb14b57790fc6

**Comment by gezhiwei8899** [comment]
image

**Comment by gezhiwei8899** [comment]
Never trigger ckp again

**Comment by gezhiwei8899** [comment]
image

**Comment by pltbkd** [comment]
@gezhiwei8899 
The commit is revert on 1.15 due to FLINK-27148. Both of the issues are now fixed on the master branch, but are not included in any 1.15 branch. I'm now trying to back port the changes to release-1.13, release-1.14 and release-1.15.

https://github.com/apache/flink/pull/20681
https://github.com/apache/flink/pull/20682
https://github.com/apache/flink/pull/20683

By the way, "Checkpoint Coordinator is suspending" is expected while the job is doing failover, and the coordinator should be restarted along with the job after the failover is done. I suppose it has nothing to do with the issue in this PR.

**Pull request merged**
